src/NHibernate/Engine/TwoPhaseLoad.cs

```diff
@@ -84,15 +85,28 @@ public static void InitializeEntity(object entity, bool readOnly, ISessionImplem
 				log.Debug("resolving associations for {0}", MessageHelper.InfoString(persister, id, session.Factory));
 
 			IType[] types = persister.PropertyTypes;
+			var collectionToResolveIndexes = new List<int>();
 			for (int i = 0; i < hydratedState.Length; i++)
 			{
 				object value = hydratedState[i];
 				if (!Equals(LazyPropertyInitializer.UnfetchedProperty, value) && !(Equals(BackrefPropertyAccessor.Unknown, value)))
 				{
+					if (types[i].IsCollectionType)
+					{
+						// Resolve them last, because they may depend on other properties if they use a property-ref
+						collectionToResolveIndexes.Add(i);
```

Comment:
The change in `CollectionType.GetKeyOfOwner` renders this change here no more mandatory. But I think it is better to remove a code path relying on `CollectionType.GetKeyOfOwner` "best effort", which the change here does.